### PR TITLE
Feature/atrayano/#169 add datalake

### PR DESCRIPTION
**Pull request by atrayano**
This PR adds a DataLake implementation in GEOS. It is zero-diff when the DataLake feature is not exercised (default), otherwise expect non-zero differences

### CODE REVIEW

**Review by gmao-rreichle — DISMISSED** [review]

**Comment by sanAkel** [comment]
https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/169#issuecomment-1069339597

**Comment by sanAkel** [comment]
@atrayano, @sdrabenh,

You may find following instructions to generate a mask for the Great Lakes and Caspian sea and (binary) SST and Sea Ice concentration helpful.

## Steps to generate a lake _mask_ and data for a given date

   In the following example, date is set to: **2022/01/01**

0. Set modules: source GEOSgcm/@env/g5_modules
1. Generate an input file `input.txt`
   - One can find `gen_input.csh` at:
     ```
     GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/
     ```
   - Change year `2020` to `2022`, i.e., line:
     https://github.com/GEOS-ESM/GMAO_Shared/blob/5b864daa4e41df596e4b385cd61da57e0f69f97b/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/gen_input.csh#L65
     to:
     ```
     set reynolds_data_loc = '/gpfsm/dnb42/projects/p17/production/GEOS5odas-5.00/RC/OBS/REYN/2022/'
     ```
   - Use `gen_input.csh 2022 1 1 OSTIA`. Attached is the generated `input.txt` if you do not _run_ `gen_input.csh` 
2. Generate data and lake mask:
   ```
   GEOSgcm/install/bin/lake_sst_sic_EIGTHdeg.x input.txt
   ```

   All goes okay, you should get:
   ```
   Data over Great Lakes and Caspian Sea
   Processing SST and ICE data from: 20220101...To... 20220102
   ...Finished!
   ```
   and three binary files: `mask.bin, Ostia_ice_20220101.bin  Ostia_sst_20220101.bin`
3. Note that these three binary files are of the same size, each contains a two-dimensional field (mask, sst or sea ice concentration),
   at a regular 1/8-deg (lat-lon) grid (date-line edge, pole edge). One can use [plot_bcs.py](https://gist.github.com/sanAkel/8845edab405c4a8ee934acacaa924a53) to plot:
   ```
   plot_bcs.py -data_path /discover/nobackup/sakella/TMP/nn -year 2022 -month 1 -day 1
   ```

[input.txt](https://github.com/GEOS-ESM/GEOSgcm_GridComp/files/8275836/input.txt)
bcs_SST_SIC_20220101

**Comment by sdrabenh** [comment]
@atrayano just checking on your timetable for generating the new raster/tile files reverting the lakes to lakes? Obviously I can't test anything further until that happens.

**Comment by atrayano** [comment]
Scott,
Thanks for your e-mail (this was a perfect reminder!).
I am planning to generate these files myself and test the PR myself first, before I pass it to you for further testing. Hopefully this won’t take too long, and I’ll be ready sometime this week…

Atanas

From: Scott Rabenhorst ***@***.***>
Reply-To: GEOS-ESM/GEOSgcm_GridComp ***@***.***>
Date: Tuesday, March 22, 2022 at 11:01 AM
To: GEOS-ESM/GEOSgcm_GridComp ***@***.***>
Cc: "Trayanov, Atanas L. (GSFC-610.1)[SCIENCE SYSTEMS AND APPLICATIONS INC]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [GEOS-ESM/GEOSgcm_GridComp] Feature/atrayano/#169 add datalake (PR #557)


@atrayano<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fatrayano&data=04%7C01%7CAtanas.L.Trayanov%40nasa.gov%7Ca65724589752482fbf0808da0c14d079%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637835580778986921%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=n69pL3Uq6Cd676pEWF57WmjdkI1YWkBNMqT4sE7V2JA%3D&reserved=0> just checking on your timetable for generating the new raster/tile files reverting the lakes to lakes? Obviously I can't test anything further until that happens.

**Comment by sdrabenh** [comment]
> Scott, Thanks for your e-mail (this was a perfect reminder!). I am planning to generate these files myself and test the PR myself first, before I pass it to you for further testing. Hopefully this won’t take too long, and I’ll be ready sometime this week… Atanas

Great, thanks. I just wanted to have a general idea of the timing. Sounds good.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 1]
```diff
@@ -1713,6 +1735,60 @@ subroutine LAKECORE(NT,RC)
...
+
+
+       allocate(SST(NT), stat=status); VERIFY_(STATUS)
+       call MAPL_ReadForcing(MAPL, 'SST', mystate%sstfile,&
+            CURRENT_TIME, SST, ON_TILES=.true., RC=STATUS)
+       VERIFY_(STATUS)
+
+       where(mystate%mask)
+          TS(:,WATER) = SST
+          TS(:,ICE)   = TICE
```
Comment (sanAkel) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1787]: `TS(:,ICE)` needs to be set from data file, just as `SST` is being set.

This needs to be _fixed_. Thanks! 🙏

**Review by sanAkel — CHANGES_REQUESTED** [review]

**Comment by sanAkel** [comment]
@sdrabenh, @atrayano please see above for `TICE` that needs to be handles just as `SST`.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 2]
```diff
@@ -1553,6 +1568,13 @@ subroutine LAKECORE(NT,RC)
    real, parameter :: LAKECAP      = (MAPL_RHOWTR*MAPL_CAPWTR*LAKEDEPTH    )
    real, parameter :: LAKEICECAP   = (MAPL_RHOWTR*MAPL_CAPWTR*LAKEICEDEPTH )
 
+   logical :: datalake
+   integer :: dlk
+   real, allocatable :: SST(:)
+   character(len=ESMF_MAXSTR) :: maskfile
+   real, parameter :: tice = MAPL_TICE - 1.8
```
[the remark this answers is not in the document]
Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1575]: @sanAkel If I remember correctly, couple of year ago I had a discussion with Andrea, and TICE was her suggestion. Are you sure the ice temperature over the lakes is resolved correctly in the observed datasets? In GEOS_DataSeaGridComp.F90 we cap the SST to TICE in a similar fashion

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 (hunk 1 above)
Comment (sanAkel) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1787]: > `TS(:,ICE)` needs to be set from data file, just as `SST` is being set.
> 
> This needs to be _fixed_. Thanks! 🙏

⬆️ is incorrect! I amend below:

```
where(mask)
 TS(:,WATER) = SST  ! from data file
  TS(:,ICE)  =  TICE ! or Tfreeze, which is the temperature where fresh water freezes, say -1.8 degC.
endwhere
```

However, I would still request that the `FR` from a dataset like the `sst.data` be also read in and we make use that `FR` where `mask = .TRUE.`

Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1787]: I totally agree, that adding FR similar to SST is the right thing to do

Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1787]: I will make the FR changes soon

Comment (sanAkel) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1787]: 🙏

**Comment by atrayano** [comment]
@sanAkel The changes you requested have been committed and pushed. Could you, pease, review them. I tested them with c12, and generating data for Scott to test them with c180 and c30

**Comment by sanAkel** [comment]
> @sanAkel The changes you requested have been committed and pushed. Could you, pease, review them. I tested them with c12, and generating data for Scott to test them with c180 and c360

Thanks @atrayano, will take a look at it later.

As for @sdrabenh's testing, you are referring to test with coupled or uncoupled model configuration? Latter may need that raster/tile file that reverts Great Lakes and Caspain to being lakes again.

**Comment by atrayano** [comment]
This up to him. He indicated desire to test both, but for now he has asked me for 2 different resolutions ( c180 and c360) uncoupled

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 3]
```diff
@@ -1773,6 +1869,7 @@ subroutine LAKECORE(NT,RC)
 ! Update Ice fraction
 !--------------------
     do I=1,NT
+       if(mystate%mask(i)) cycle
```
Comment (sanAkel) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1872]: @atrayano,
probably it should be:
```
mystate%mask
```
**without** `(i)` ??

**Review by sanAkel — COMMENTED** [review]
@atrayano thanks for adding the ice fraction. Please see one question ⬆️ reg `mystate%mask(i)`

**Comment by sdrabenh** [comment]
> This is up to him. He indicated desire to test both, but for now he has asked me for 2 different resolutions ( c180 and c360) uncoupled. Of course, my new tile file reflects the reverted "new" lakes

Thanks @atrayano. @sanAkel these are for the uncoupled tests just to test if everything is behaving as expected and within roundoff. However, when we meet this week I'd like to discuss what you recommend for a coupled test.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 (hunk 3 above)
Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1872]: @sanAkel The "i" is needed at the moment. This block needs to be revisited or handled differently altogether, but not as a part of this PR

**Comment by atrayano** [comment]
Pinging @sanAkel to review the PR

**Review by sanAkel — DISMISSED** [review]

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 (hunk 3 above)
Comment (sanAkel) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1872]: @atrayano, I don't understand why not!

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 4]
```diff
@@ -1713,6 +1737,81 @@ subroutine LAKECORE(NT,RC)
...
+       allocate(mystate%mask(NT), stat=status); VERIFY_(STATUS)
+       mystate%mask = .false.
+    end if
+           
+    if (datalake) then
+
+        ! next section is done only once. We do it here since the Initalize
+        ! method of this component defaults to MAPL_GenericInitialize
+        
+        if (.not. mystate%InitDone) then
```
Comment (sdrabenh) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1763]: @atrayano don't we have to update the data as the run goes on?

Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1763]: @sdrabenh We keep only the mask (i.e. mapping which lake tiles we use to apply dataLake SSTs) in the private state. The mask does not change during the run, only the SST and the fractions of the lake ice

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 5]
```diff
@@ -1713,6 +1737,81 @@ subroutine LAKECORE(NT,RC)
...
+       allocate(DATA_SST(NT), DATA_FR(NT), stat=status); VERIFY_(STATUS)
+       call MAPL_ReadForcing(MAPL, 'SST', mystate%sstfile,&
+            CURRENT_TIME, DATA_SST, ON_TILES=.true., RC=STATUS)
+       VERIFY_(STATUS)
+       call MAPL_ReadForcing(MAPL, 'FR', mystate%dataFrtFile,&
+            CURRENT_TIME, DATA_FR, ON_TILES=.true., RC=STATUS)
+       VERIFY_(STATUS)
+
+       do I=1,NT
+          if(mystate%mask(i)) then
```
Comment (sdrabenh) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1798]: @atrayano are we sure this condition is actually being entered for our mask?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 6]
```diff
@@ -1749,7 +1848,7 @@ subroutine LAKECORE(NT,RC)
 ! Update surface temperature and moisture
 !----------------------------------------
 
-       TS(:,N) = TS(:,N) + DTS
+       where(.not.mystate%mask) TS(:,N) = TS(:,N) + DTS
```
Comment (sdrabenh) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1851]: @atrayano shouldn't we loop through `I=1,NT` ?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 (hunk 5 above)
Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1798]: I am almost sure that we enter the block above. I am setting a debugging experiment right now...

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 (hunk 6 above)
Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:1851]: The loop over NT is implicit (i.e. the tile index is the first index in TS. But we also loop over ice and water (N index)

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90 [hunk 7]
```diff
@@ -1919,6 +2022,54 @@ end subroutine ALBLAKEICE
 
 end subroutine RUN2
 
+subroutine DataLakeReadMask(mapl, msk, maskfile, rc)
+  implicit none
+  ! arguments
+  type (MAPL_MetaComp), pointer  :: MAPL
+  logical, intent(INOUT)           :: msk(:)
```
[the remark this answers is not in the document]
Comment (atrayano) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlake_GridComp/GEOS_LakeGridComp.F90:2029]: No (but this is why we had confusion with the different lake mask files). The local variable is logical, but it is set inside DatLakeReadMask (where we read a "float" and convert it to logical)

**Comment by sanAkel** [comment]
@tclune, @atrayano I just looked at this PR with @sdrabenh.

The code seems to have many bugs and superfluous logic.

We recommend this PR be closed, code be **cleanly** rewritten for our consideration.

Thank you.

**Review by gmao-rreichle — APPROVED** [review]

**Review by sdrabenh — APPROVED** [review]

**Review by sdrabenh — APPROVED** [review]

**Pull request merged**